README.pod
=pod

=head1 NAME

schedule.pl - Print duty schedule table

=head1 SYNOPSIS

  schedule.pl -f <from date> -t <to date> -o <op list> --duties <duties> -out <tt|yaml|json|ascii> -h 
  schedule.pl -in <yaml|json> -out <tt|yaml|json|ascii> < schedule.json

  Options:
    -f from date
    -t to date
    -o comma separated operators list
    -d duties
    -out <tt=<template path>|yaml|json>
    -in <json|yaml>
    -h help

=head1 OPTIONS

=over 4

=item B<--from|-f>

Schedule start date in format YYYY-MM-DD or 'now', default now

=item B<--to|-t>

Schedule end date in format YYYY-MM-DD or +<num><d|w|m>, default +3weeks

=item B<---operators|-o>

Comma separated duty operators list, eg. -o ivanov,petrov,sidorov

=item B<--duties|-d>

Duties in format duty=days, eg.  -d Support=1,Release=2

=item B<--output|-out>

Choose output format <tt=<template path>|yaml|json>, default -out ascii

=item B<--input|-in>

Read inpout from stdin in format <json|yaml>, insted generation

=item B<--country_code|-cc>

Dtermine holidays via Date::Holidays
(specific country module must be installed)
                      
=item B<--today>

Day to pass to template as 'today' in format YYYY-MM-DD, default today

=item B<-h>

This message

=back

=head1 DESCRIPTION

B<scheduel.pl> is program to to generate and draw duty schedule table in several formats

  $ perl scripts/schedule.pl -o Mike,John,Bob -out ascii -f 2018-08-13 -t +7d -d "support=1,tasks=2"

  .-------------------------------------------------------------------------------.
  | Month/Day |  Aug/13 |  Aug/14 |  Aug/15 |  Aug/16 |  Aug/17 | Aug/18 | Aug/19 |
  +-----------+---------+---------+---------+---------+---------+--------+--------+
  | Mike      | support |         | support |         | tasks   |        |        |
  | John      | tasks   | tasks   |         | support |         |        |        |
  | Bob       |         | support | tasks   | tasks   | support |        |        |
  '-----------+---------+---------+---------+---------+---------+--------+--------'

=head1 AUTHOR

Mike Pruzhanskiy <pru.mike@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2018 Mike Pruzhanskiy <pru.mike@gmail.com>

This is free software; you can redistribute it and/or modify it under the same terms 
as the Perl 5 programming language system itself.

=head1 SEE ALSO

L<App::Schedule::Generate>



Changes

0.01  2018-08-12
  - initial release



scripts/send_to_slack.pl
#!/usr/bin/env perl

use strict;
use warnings;
use FindBin qw/$Bin/;
use lib qq{$Bin/../lib};
use Getopt::Long;
use Pod::Usage qw(pod2usage);
use App::Schedule::Generate;
use LWP::UserAgent;
use URL::Encode qw/url_encode/;

our @TT_PARAMS = (ABSOLUTE => 1, RELATIVE => 1);

sub HELP_MESSAGE(;$);

my $tt = q[templates/slack.tt];
my ($h, $url, $input, $dump, $date, $link, $title);
my @inputs = qw/json yaml/;
my $dt = DateTime->now();

GetOptions(
    "input|in=s"    => \$input,
    "dump|d"        => \$dump,
    "url|u=s"       => \$url,
    "tempalte|tt=s" => \$tt,
    "help|h"        => \$h,
    "date=s"        => \$date,
    "link|l=s"      => \$link,
    "title|t=s"     => \$title,
);
HELP_MESSAGE() if $h;
HELP_MESSAGE "URL not defined"          unless $url;
HELP_MESSAGE "Input format not defined" unless $input;
HELP_MESSAGE "Template not defined"     unless -f $tt;

if ($date) {
    if ($date =~ /(\d{4})-(\d\d)-(\d\d)/) {
        $dt = DateTime->new(
            year  => $1,
            month => $2,
            day   => $3,
        );
    } else {
        HELP_MESSAGE "wrong --date <date> format";
    }
}

my $data = do { local $/; <> };
my $table = App::Schedule::Generate->load($input => $data);

HELP_MESSAGE "Bad input [$input], MUST be one of {@inputs}" unless $table;

send_slack_notification($tt, $dt, $url, $table, $dump);

sub send_slack_notification {
    my ($tt, $dt, $slack_webhook_url, $table, $dump) = @_;

    my ($op) = App::Schedule::Generate->find_in_table($dt, $table);
    unless ($op) {
        warn "No one is on duty at $dt\n";
        return;
    }

    my $content;
    my $template = Template->new(@TT_PARAMS);
    $template->process($tt, { op => lc($op) }, \$content) || die $template->error();

    my $ua  = LWP::UserAgent->new();
    my $req = HTTP::Request->new(
        POST => $slack_webhook_url,
        ['Content-Type' => 'application/x-www-form-urlencoded'],
        'payload=' . url_encode($content)
    );
    if ($dump) {
        print " ================ message ================\n";
        print $content;
        print " ============= http request ==============\n";
        print $req->as_string;
        print " =========================================\n";
    } else {
        my $res = $ua->request($req);
        unless ($res->is_success) {
            warn 'Request failed: ', $res->status_line, "\n";
        }
    }
}

sub HELP_MESSAGE(;$) {
    my $err_msg = shift;

    pod2usage({
            -msg     => "\n[ERROR]: $err_msg\n",
            -vebose  => 0,
            -exitval => 2,
        }
    ) if $err_msg;

    pod2usage({ -verbose => 2 });

}

__END__

=pod

=head1 NAME

send_to_slack.pl - Find today operator and send to slack

=head1 SYNOPSIS

  send_to_slack.pl -u <url> -in <yaml|json> -h < schedule.json

=head1 OPTIONS

=over 4

=item B<--url|-u>

Slack webhook url

=item B<--template|-tt>

Slack json message template, default templates/slack.tt

=item B<--input|-in>

Input message format, one of json, yaml

=item B<--dump|-d>

Dump message to stdout insted of POST to slack

=item B<--link|-l>

Pass to template link to schedule

=item B<--title|-t>

Pass to template message title 

=item B<-h>

This message

=back

=head1 AUTHOR

Mike Pruzhanskiy <pru.mike@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2018 Mike Pruzhanskiy <pru.mike@gmail.com>

This is free software; you can redistribute it and/or modify it under the same terms 
as the Perl 5 programming language system itself.

=head1 SEE ALSO

L<App::Schedule::Generate>, B<schedule.pl>

=cut



lib/App/Schedule/Generate.pm
package App::Schedule::Generate;

use strict;
use warnings;
use feature qw/state/;

use DateTime;
use Carp q/croak/;
use List::Util qw/pairmap/;
use Template;

our $VERSION = "0.1";

use Class::Accessor::Fast qw/moose-like/;

has fdt => (is => 'ro', isa => 'DateTime');
has tdt => (is => 'ro', isa => 'DateTime');

use constant {
    MON => 1,
    TUE => 2,
    WED => 3,
    THU => 4,
    FRI => 5,
    SAT => 6,
    SUN => 7,
};

use constant {
    DUTY_DAYS  => 0,
    DUTY_OPS   => 1,
    CURRENT_OP => 0,
    WORKDAY    => 0,
    HOLIDAY    => 1,
    ON         => 1,
    OFF        => 0,
};

our @TT_PARAMS = (ABSOLUTE => 1, RELATIVE => 1);

sub new {
    my $class = shift;
    my $p     = shift;

    for (qw/from_date to_date/) {
        croak "Missing required parameter '$_'" unless ($p->{$_});
    }
    if ($p->{holidays} and ref($p->{holidays}) ne 'HASH') {
        croak 'Bad $p->{holidays} parameter, must be hashref';
    }

    my %weekend_conf;
    if ($p->{holidays}) {
        for (qw/or_weekend no_weekend/) {
            if (exists $p->{holidays}{$_}) {
                $weekend_conf{$_} = delete $p->{holidays}{$_};
            }
        }
    }
    if ($p->{holidays}) {
        if (not _load_lib('Date::Holidays', 0)) {
            warn 'Date::Holidays not installed, but holidays is configured';
            delete $p->{holidays};
        }
    }
    if ($p->{holidays} and !grep(exists $p->{holidays}{$_}, qw/countries state regions/)) {
        warn '$p->{holidays} do not properly configured, countries or state or regions should be defined';
        delete $p->{holidays};
    }

    if (not $p->{duties} or ref($p->{duties}) ne 'ARRAY') {
        croak 'duties not defined';
    } else {
        if (@{ $p->{duties} } % 2 == 1) {
            croak 'duties MUST be even sized array in <duty => days> format';
        } else {
            $p->{duties} = [pairmap { [$a => $b] } @{ $p->{duties} }];
        }
    }

    my $self = {
        fdt           => delete $p->{from_date},
        tdt           => delete $p->{to_date},
        holidays_conf => delete $p->{holidays},
        %weekend_conf,
        head => [[Month => 'month_abbr'], [Day => 'day'], [Wday => 'day_abbr'],],
        duties             => delete $p->{duties},
        operators          => delete $p->{operators},
        duties_at_holidays => {},
        duties_at_workdays => {},
        %{$p}
    };
    bless $self, $class;
}

sub make_schedule {
    my $self = shift;
    my @schedule;
    push @schedule, @{ $self->_make_head() };
    push @schedule, @{ $self->_make_body() };
    return \@schedule;
}

sub _make_head {
    my $self = shift;
    [map { [$_->[0], _datetime_method($_->[1] => $self->fdt, $self->tdt)] } @{ $self->{head} }];
}

sub _is_holiday {
    my $self        = shift;
    my $dt          = shift;
    my $is_weekends = _check_weekend($dt);
    if ($self->{holidays_conf}) {
        my $is_holiday = _check_by_date_holidays($dt, $self->{holidays_conf});
        if ($self->{or_weekend}) {
            return $is_weekends || $is_holiday;
        } elsif ($self->{no_weekend}) {
            return $is_weekends ? 0 : $is_holiday;
        } else {
            return $is_holiday;
        }
    } elsif ($self->{no_weekend}) {
        return 0;
    } else {
        return $is_weekends;
    }
}

sub _check_by_date_holidays {
    my ($dt, $conf) = @_;
    my $is_russian_holidays = 0;
    if (grep(/^ru$/i, @{ $conf->{countries} })) {
        _load_lib('Date::Holidays::RU');
        $is_russian_holidays = !Date::Holidays::RU::is_business_day(map { $dt->$_ } qw/year month day/);
        if (1 == @{ $conf->{countries} }) {
            return $is_russian_holidays ? 1 : 0;
        }
    }

    my $holidays = Date::Holidays->is_holiday(%{$conf}, map { $_ => $dt->$_ } qw/year day month/,) || {};

    return grep ($_, values %$holidays) || $is_russian_holidays;
}

sub _check_weekend {
    my $dt = shift;
    grep($dt->day_of_week == $_, SAT, SUN) ? 1 : 0;
}

sub _datetime_method {
    my ($m, $f, $t) = @_;
    my @res;
    for (my $f = $f->clone; $f < $t; $f->add(days => 1)) {
        push @res, $f->$m;
    }
    return @res;
}

sub _rotate {
    my $list = shift;
    push @{$list}, shift @{$list};
}

sub _make_body {
    my $self      = shift;
    my $duties    = $self->{duties} || croak '$duties not defined';
    my $operators = $self->{operators} || croak '$operators not defined';
    croak '$duties is empty'    unless (@$duties);
    croak '$operators is empty' unless (@$operators);

    for (@$duties) {
        die "Wrong duties format [@$_], must be ['duties name' => 'duties days']" if (@$_ != 2);
        die "Wrong duties format [@$_], 'duties days' must be positive]" if ($_->[1] < 0);
    }

    my $table = [];
    for (my $i = 0; $i < @$operators; $i++) {
        push @{ $table->[$i] }, $operators->[$i];
    }

    my $duty_queue    = {};
    my @operators_tmp = @$operators;
    for (@$duties) {
        my ($m, $dd) = ($_->[0], $_->[1]);
        $duty_queue->{$m} = [$dd, [@operators_tmp]];
        _rotate(\@operators_tmp);
    }

    my $duty_by_holidays = {};
    for my $d (@$duties) {
        $duty_by_holidays->{ $d->[0] }[WORKDAY] = ON;
        $duty_by_holidays->{ $d->[0] }[HOLIDAY] = OFF;
    }
    for (qw/workdays holidays/) {
        my $h = qq[duties_at_${_}];
        my $v = /workdays/ ? 0 : 1;
        for my $d (keys %{ $self->{$h} }) {
            $duty_by_holidays->{$d}[$v] = $self->{$h}{$d};
        }
    }

    for (my $f = $self->fdt->clone; $f < $self->tdt; $f->add(days => 1)) {

        my $is_holiday = $self->_is_holiday($f);

      OP: for (my $i = 0; $i < @$operators; $i++) {
            my $op = $operators->[$i];

            for my $d (keys %$duty_queue) {
                unless ($duty_by_holidays->{$d}[$is_holiday]) {
                    next;
                }
                if ($duty_queue->{$d}[DUTY_OPS][CURRENT_OP] eq $op) {
                    push @{ $table->[$i] }, $d;
                    $duty_queue->{$d}[DUTY_DAYS]--;
                    next OP;
                }
            }
            push @{ $table->[$i] }, '';
        }

        for my $d (keys %$duty_queue) {
            unless ($duty_queue->{$d}[DUTY_DAYS]) {
                _rotate($duty_queue->{$d}[DUTY_OPS]);
                my $op = $duty_queue->{$d}[DUTY_OPS][CURRENT_OP];
                for (keys %$duty_queue) {
                    unless ($duty_by_holidays->{$_}[$is_holiday]) {
                        next;
                    }
                    if ($duty_queue->{$_}[DUTY_DAYS] and $op eq $duty_queue->{$_}[DUTY_OPS][CURRENT_OP]) {
                        _rotate($duty_queue->{$d}[DUTY_OPS]);
                    }
                }
                $duty_queue->{$d}[DUTY_DAYS] = (grep { $d eq $_->[0] } @$duties)[0][1];
            }
        }
    }

    return $table;
}

sub draw_tt {
    my $class    = shift;
    my $template = shift || croak 'Template not defined';
    my $table    = shift || croak 'table not defined';
    my $vars     = shift || {};
    my $tt       = Template->new(@TT_PARAMS);
    my $output;
    $tt->process($template, { %$vars, table => $table }, \$output) || croak $tt->error();
    $output;
}

sub draw_ascii {
    my $class = shift;
    my $table = shift || croak 'table not defined';
    my $args  = shift || {};
    _load_lib('Text::ASCIITable');
    my $t = Text::ASCIITable->new({ alignHeadRow => 'center', %$args });
    $t->setCols([map { join '/', $table->[0][$_], $table->[1][$_] } 0 .. $#{ $table->[0] }]);
    $t->addRow(@{ $table->[$_] }) for 3 .. $#{$table};
    "$t";
}

sub _load_lib {
    my $lib = shift;
    my $required = shift // 1;
    state $cache = {};
    $cache->{$lib} = eval "use $lib;1" if not exists $cache->{$lib};
    !$cache->{$lib} and $required and die "Can't load $lib library";
    $cache->{$lib};
}

sub load {
    my $class = shift;
    my $input = shift;
    my $data  = shift;
    my $table;
  INPUT:
    for ($input) {
        if (/^json$/) {
            _load_lib('JSON');
            $table = JSON->new->decode($data);
            last INPUT;
        }
        if (/^yaml$/) {
            _load_lib('YAML');
            $table = YAML::Load($data);
            last INPUT;
        }
    }
    return $table;
}

sub find_in_table {
    my $class = shift;
    my $dt    = shift || croak 'DateTime not defined';
    my $table = shift || croak 'Table not defined';
    my ($mon, $day) = ($dt->month_abbr, $dt->day);

    my (@col, $col, $op);
    for (my $i = 1; $i < @{ $table->[0] }; $i++) {
        push @col, $i if $table->[0][$i] eq $mon;
    }
    for (@col) {
        if ($table->[1][$_] == $day) {
            $col = $_;
            last;
        }
    }
    unless ($col) {
        return;
    }

    for (my $i = 3; $i < @$table; $i++) {
        if ($table->[$i][$col]) {
            $op = $table->[$i][0];
            last;
        }
    }
    unless ($op) {
        return;
    }
    return ($op, $col);
}

1;

__END__

=pod

=head1 NAME

App::Schedule::Generate - Simple schedule table generator

=head1 SYNOPSIS

  use App::Schedule::Generate;

  my $app = App::Schedule::Generate->new({
      from_date => DateTime->now,
      to_date => DateTime->now->add(weeks => 1),
      duties => ['carry ring' => 1, 'find food' => 1],
      operators => [qw/Frodo Sam Golum/],
  });

  my $table = $app->make_schedule();
  print App::Schedule::Generate->draw_ascii($table);

=head1 DESCRIPTION

Simple module to generate and draw duty schedule table, looking like:

  .-------------------------------------------------------------------------------.
  | Month/Day |  Aug/13 |  Aug/14 |  Aug/15 |  Aug/16 |  Aug/17 | Aug/18 | Aug/19 |
  +-----------+---------+---------+---------+---------+---------+--------+--------+
  | Mike      | support |         | support |         | tasks   |        |        |
  | John      | tasks   | tasks   |         | support |         |        |        |
  | Bob       |         | support | tasks   | tasks   | support |        |        |
  '-----------+---------+---------+---------+---------+---------+--------+--------'

=head1 Methods

=head2 new($href)

Define schedule parameters

=head3 Required

=over 4

=item from_date => Date::Time

Schedule start date

=item to_date => Date::Time

Schedule end date

=item duties => $arrayref

Even sized duities list in format B<duty =E<gt> days>, e.g.

  duties => [support => 1, tasks => 2]

=item operators => $arrayref

Operators list, e.g.

  operators => [qw/Mike John Bob/]

=back

=head3 Optional

=over 4

=item holidays => $hashref

Configure holidays

  holidays => {
      or_weekend => 1,
      countries  => 'ru'
  }

  holidays => {
      no_weekend => 0,
      countries  => 'us'
      state      => ...
  }

=back

=over 8

=item or_weekend

Weekend is holiday

=item no_weekend

Weekend is no holiday

=item countries or state or regions

Define holidays via L<Date::Holidays> module.

=back

=head2 make_schedule()

  $table = $app->make_schedule()

Return hashref table with schedule in format:

=head2 draw_tt($tt,$table,$args)

  $output = App::Schedule::Generate->draw_tt($tt,$table,{});

Draw table B<$table> via template <$tt>

=head2 draw_ascii($table,$args)

  $output = App::Schedule::Generate->draw_ascii($table);

Draw table B<$table> via L<Text::ASCIITable>
L<Text::ASCIITable> module must be installed;

=head2 load($input,$data)

  $input = 'json'; # yaml
  $table = App::Schedule::Generate->load($input => $data);

Conversion beetwen yaml or json to perl data
Relevant module (JSON,YAML) must be installed

=head2 find_in_table($date_time,$table)

  $date = Date::Time->now;
  ($operator, $column) = App::Schedule::Generate->find_in_table($date,$table);

Find current operator on presented day in schedule table.

=head1 AUTHOR

Mike Pruzhanskiy <pru.mike@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2018 Mike Pruzhanskiy <pru.mike@gmail.com>

This is free software; you can redistribute it and/or modify it under the same terms 
as the Perl 5 programming language system itself.

=cut



scripts/schedule.pl
#!/usr/bin/env perl

use strict;
use warnings;
use FindBin qw/$Bin/;
use lib qq{$Bin/../lib};
use Getopt::Long;
use Pod::Usage qw(pod2usage);
use App::Schedule::Generate;

sub HELP_MESSAGE(;$);

my $from = 'now';
my $to   = '+3w';
my @operators;
my @duties;
my $h;
my $cc;
my $tt;
my @inputs = qw/json yaml/;
my $input;
my @outputs = qw/ascii tt json yaml/;
my $output  = $outputs[0];
my $table;
my $today = 'now';
my $todaydt;

GetOptions(
    "help|h"          => \$h,
    "from|f=s"        => \$from,
    "to|t=s"          => \$to,
    "ops|o=s"         => \@operators,
    "duties|d=s"      => \@duties,
    "country_code|cc" => \$cc,
    "input|in=s"      => \$input,
    "output|out=s"    => \$output,
    "today=s"         => \$today,
);
HELP_MESSAGE() if $h;

if ($output && $output =~ /^(tt)=(.+)/) {
    ($output, $tt) = ($1, $2);
}
HELP_MESSAGE "Bad output [$output], MUST be one of {@outputs}" if not grep($_ eq $output, @outputs);
HELP_MESSAGE "Template not defined" if $output eq 'tt' and not -f $tt;

check_date(today => $today, \$todaydt);

if (not $input) {

    @operators = split(/,/, join(',', @operators));
    @duties    = split(/,/, join(',', @duties));

    {
        my @tmp;
        for (@duties) {
            my ($d, $n) = split /=/, $_, 2;
            $n ||= 1;
            push @tmp, $d, $n;
        }
        @duties = @tmp;
        unless (@duties) {
            @duties = qw/D 1/;
        }
    }

    HELP_MESSAGE "Operators not defined" unless (@operators);

    my ($fdt, $tdt);

    check_date(from => $from, \$fdt);

    if ($to =~ /(\d{4})-(\d\d)-(\d\d)/) {
        $tdt = DateTime->new(
            year  => $1,
            month => $2,
            day   => $3,
        );
    } elsif ($to =~ m{\+\s*(\d+)?\s*(w(?:eek)?|d(?:ay)?|m(?:onth)?)s?}) {
        my $num = $1 || 1;
        my $dim = substr($2, 0, 1);
        my $add = $dim eq 'w' ? 'weeks' : $dim eq 'm' ? 'months' : 'days';
        $tdt = $fdt->clone->add($add => $num);
    } else {
        HELP_MESSAGE "wrong --to <date> format";
    }

    my $holidays;
    if ($cc) {
        $holidays = { countries => [$cc], };
    }

    my $app = App::Schedule::Generate->new({
            from_date => $fdt,
            to_date   => $tdt,
            duties    => \@duties,
            operators => \@operators,
            holidays  => $holidays,
        }
    );

    $table = $app->make_schedule()

} else {

    my $data = join '', <>;

    $table = App::Schedule::Generate->load($input => $data);
    HELP_MESSAGE "Bad input [$input], MUST be one of {@inputs}" unless $table;

}

OUTPUT:
for ($output) {
    if (/^tt$/) {
        my ($op, $tcn) = App::Schedule::Generate->find_in_table($todaydt, $table);
        print App::Schedule::Generate->draw_tt($tt, $table, { operator => $op, today_col_num => $tcn });
        last OUTPUT;
    }
    if (/^ascii$/) {
        print App::Schedule::Generate->draw_ascii($table);
        last OUTPUT;
    }
    if (/^json$/) {
        eval "use JSON;1" or die "Can't load JSON library";
        print JSON->new->encode($table);
        last OUTPUT;
    }
    if (/^yaml$/) {
        eval "use YAML;1" or die "Can't load YAML library";
        print YAML::Dump($table);
        last OUTPUT;
    }
}

sub check_date {
    my ($name, $text, $dt) = @_;
    if ($text =~ /(\d{4})-(\d\d)-(\d\d)/) {
        $$dt = DateTime->new(
            year  => $1,
            month => $2,
            day   => $3,
        );
    } elsif ($text eq 'now') {
        $$dt = DateTime->now;
    } else {
        HELP_MESSAGE "wrong --$name <$text> format";
    }
    return;
}

sub HELP_MESSAGE(;$) {
    my $err_msg = shift;

    pod2usage({
            -msg     => "\n[ERROR]: $err_msg\n",
            -vebose  => 0,
            -exitval => 2,
        }
    ) if $err_msg;

    pod2usage({ -verbose => 2 });
}

__END__

=pod

=head1 NAME

schedule.pl - Print duty schedule table

=head1 SYNOPSIS

  schedule.pl -f <from date> -t <to date> -o <op list> --duties <duties> -out <tt|yaml|json|ascii> -h 
  schedule.pl -in <yaml|json> -out <tt|yaml|json|ascii> < schedule.json

  Options:
    -f from date
    -t to date
    -o comma separated operators list
    -d duties
    -out <tt=<template path>|yaml|json>
    -in <json|yaml>
    -h help

=head1 OPTIONS

=over 4

=item B<--from|-f>

Schedule start date in format YYYY-MM-DD or 'now', default now

=item B<--to|-t>

Schedule end date in format YYYY-MM-DD or +<num><d|w|m>, default +3weeks

=item B<---operators|-o>

Comma separated duty operators list, eg. -o ivanov,petrov,sidorov

=item B<--duties|-d>

Duties in format duty=days, eg.  -d Support=1,Release=2

=item B<--output|-out>

Choose output format <tt=<template path>|yaml|json>, default -out ascii

=item B<--input|-in>

Read inpout from stdin in format <json|yaml>, insted generation

=item B<--country_code|-cc>

Dtermine holidays via Date::Holidays
(specific country module must be installed)
                      
=item B<--today>

Day to pass to template as 'today' in format YYYY-MM-DD, default today

=item B<-h>

This message

=back

=head1 DESCRIPTION

B<scheduel.pl> is program to to generate and draw duty schedule table in several formats

  $ perl scripts/schedule.pl -o Mike,John,Bob -out ascii -f 2018-08-13 -t +7d -d "support=1,tasks=2"

  .-------------------------------------------------------------------------------.
  | Month/Day |  Aug/13 |  Aug/14 |  Aug/15 |  Aug/16 |  Aug/17 | Aug/18 | Aug/19 |
  +-----------+---------+---------+---------+---------+---------+--------+--------+
  | Mike      | support |         | support |         | tasks   |        |        |
  | John      | tasks   | tasks   |         | support |         |        |        |
  | Bob       |         | support | tasks   | tasks   | support |        |        |
  '-----------+---------+---------+---------+---------+---------+--------+--------'

=head1 AUTHOR

Mike Pruzhanskiy <pru.mike@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2018 Mike Pruzhanskiy <pru.mike@gmail.com>

This is free software; you can redistribute it and/or modify it under the same terms 
as the Perl 5 programming language system itself.

=head1 SEE ALSO

L<App::Schedule::Generate>

=cut

